### PR TITLE
test: [VRD-725] Add batch predict property tests

### PR DESCRIPTION
**Pull request by ghost**
## Impact and Context
Add a simple property test to generate a whole bunch of dataframes.

## Testing
- [X] Unit test
- [ ] Deployed to dev env
- [ ] Other (explain) 

## Reverting
- [ ] Contains Migration - _Do Not Revert_

### CODE REVIEW

File: client/verta/tests/unit_tests/test_deployed_model.py [hunk 1]
```diff
@@ -456,26 +461,84 @@ def test_batch_predict_with_batches_and_indexes(mocked_responses) -> None:
...
+
+@st.composite
+def generate_data(draw):
+    """ Return a dict that represents a dataframe. Generates ints, floats, and strings."""
+    num_rows = draw(st.integers(min_value=50, max_value=200))
+    num_cols = draw(st.integers(min_value=1, max_value=10))
+    col_names = draw(st.lists(st.text(), max_size=num_cols, min_size=num_cols, unique=True))
+    data = {}
+    for name in col_names:
+        type_probability = draw(st.floats(min_value=0, max_value=1))
```
Comment (liuverta) [inline client/verta/tests/unit_tests/test_deployed_model.py]: I think we should use [`random.random()`](https://docs.python.org/3/library/random.html#random.random). `st.floats()` as a strategy has shrinking behavior (i.e. it does something smart to narrow down cases, and isn't necessarily a uniform distribution over time) which wouldn't make sense if we just simply want to roll a die to pick a column type.

File: client/verta/tests/unit_tests/test_deployed_model.py [hunk 2]
```diff
@@ -456,26 +461,84 @@ def test_batch_predict_with_batches_and_indexes(mocked_responses) -> None:
...
     input_df = pd.DataFrame({"a": [1, 2, 3, 4, 5], "b": [11, 12, 13, 14, 15]}, index=["A", "B", "C", "D", "E"])
     prediction_df = dm.batch_predict(input_df, 1)
     expected_final_df = pd.concat(expected_df_list)
     pd.testing.assert_frame_equal(expected_final_df, prediction_df)
+
+
+@st.composite
+def generate_data(draw):
+    """ Return a dict that represents a dataframe. Generates ints, floats, and strings."""
+    num_rows = draw(st.integers(min_value=50, max_value=200))
```
Comment (liuverta) [inline client/verta/tests/unit_tests/test_deployed_model.py]: Is there a reason not to go with

```suggestion
    num_rows = draw(st.integers(min_value=1, max_value=200))
```

? I would imagine we'd want our functionality to work with DataFrames down to 1 row.

File: client/verta/tests/unit_tests/test_deployed_model.py [hunk 3]
```diff
@@ -456,26 +461,84 @@ def test_batch_predict_with_batches_and_indexes(mocked_responses) -> None:
...
+    col_names = draw(st.lists(st.text(), max_size=num_cols, min_size=num_cols, unique=True))
+    data = {}
+    for name in col_names:
+        type_probability = draw(st.floats(min_value=0, max_value=1))
+        if type_probability <= 0.3:
+            col = draw(st.lists(st.integers(), max_size=num_rows, min_size=num_rows))
+        elif type_probability <= 0.6:
+            col = draw(st.lists(st.floats(), max_size=num_rows, min_size=num_rows))
+        else:
+            col = draw(st.lists(st.text(), max_size=num_rows, min_size=num_rows))
```
Comment (liuverta) [inline client/verta/tests/unit_tests/test_deployed_model.py]: Slightly less code duplication:

```suggestion
        if type_probability <= 0.3:
            col_values = st.integers()
        elif type_probability <= 0.6:
            col_values = st.floats()
        else:
            col_values = st.text()
        col = draw(st.lists(col_values, max_size=num_rows, min_size=num_rows))
```

File: client/verta/tests/unit_tests/test_deployed_model.py [hunk 4]
```diff
@@ -456,26 +461,84 @@ def test_batch_predict_with_batches_and_indexes(mocked_responses) -> None:
...
+        data[name] = col
+
+    out_dict = {"data": data}
+    index_probability = draw(st.floats(min_value=0, max_value=1))
```
Comment (liuverta) [inline client/verta/tests/unit_tests/test_deployed_model.py]: Ditto on my above comment about `random.random()`

File: client/verta/tests/unit_tests/test_deployed_model.py [hunk 5]
```diff
@@ -456,26 +461,84 @@ def test_batch_predict_with_batches_and_indexes(mocked_responses) -> None:
...
+    if index_probability <= 0.5:
+        index = draw(st.lists(st.text(), max_size=num_rows, min_size=num_rows))
+        out_dict["index"] = index
+    return out_dict
+
+
+@given(json_df=generate_data(), batch_size=st.integers(min_value=5, max_value=50))
+def test_batch(json_df, batch_size) -> None:
+    """ Test that the batch_predict method works with a variety of inputs. """
+    with responses.RequestsMock() as rsps:
```
Comment (liuverta) [inline client/verta/tests/unit_tests/test_deployed_model.py:527]: Could we use the `mocked_responses` fixture instead? Or does it not work?

File: client/verta/tests/unit_tests/test_deployed_model.py [hunk 6]
```diff
@@ -456,26 +461,84 @@ def test_batch_predict_with_batches_and_indexes(mocked_responses) -> None:
...
+
+    out_dict = {"data": data}
+    index_probability = draw(st.floats(min_value=0, max_value=1))
+    if index_probability <= 0.5:
+        index = draw(st.lists(st.text(), max_size=num_rows, min_size=num_rows))
+        out_dict["index"] = index
+    return out_dict
+
+
+@given(json_df=generate_data(), batch_size=st.integers(min_value=5, max_value=50))
```
Comment (liuverta) [inline client/verta/tests/unit_tests/test_deployed_model.py]: How would you feel about going down to 1 here, too?

```suggestion
@given(json_df=generate_data(), batch_size=st.integers(min_value=1, max_value=50))
```

**Review by liuverta — DISMISSED** [review]

File: client/verta/tests/unit_tests/test_deployed_model.py (hunk 5 above)
Comment (ghost) [reply client/verta/tests/unit_tests/test_deployed_model.py:527]: depressingly, hypothesis `given`s and pytest `fixture`s don't play well together :(

File: client/verta/tests/unit_tests/test_deployed_model.py [hunk 7]
```diff
@@ -456,26 +464,86 @@ def test_batch_predict_with_batches_and_indexes(mocked_responses) -> None:
...
+@st.composite
+def generate_data(draw):
+    """ Return a dict that represents a dataframe. Generates ints, floats, and strings."""
+    num_rows = draw(st.integers(min_value=1, max_value=200))
+    num_cols = draw(st.integers(min_value=1, max_value=10))
+    col_names = draw(st.lists(st.text(), max_size=num_cols, min_size=num_cols, unique=True))
+    data = {}
+    for name in col_names:
+
+        type_probability = utils.gen_probability()
```
Comment (ewagner-verta) [inline client/verta/tests/unit_tests/test_deployed_model.py:506]: Have you added a function `gen_probability()` to `tests.utils` that is not included in this branch?

This test fails for me with `AttributeError: module 'tests.utils' has no attribute 'gen_probability'`

**Review by ewagner-verta — APPROVED** [review]

**Pull request merged**